component/discovery/kubernetes_crds/kubernetes_crds.go

```diff
@@ -0,0 +1,336 @@
+package kubernetes_crds
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/component/prometheus"
+	"github.com/grafana/agent/pkg/util/k8sfs"
+	promop "github.com/prometheus-operator/prometheus-operator/pkg/client/informers/externalversions"
+	promCommonConfig "github.com/prometheus/common/config"
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"github.com/prometheus/prometheus/scrape"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+
+	v1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/prometheus-operator/prometheus-operator/pkg/client/versioned"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name:    "prometheus.crds",
+		Args:    Config{},
+		Exports: struct{}{},
+
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args)
+		},
+	})
+}
+
+type Component struct {
+	opts   component.Options
+	config *Config
+
+	onUpdate chan struct{}
+	mut      sync.Mutex
+}
+
+// crdManager is all of the fields required to run the component.
+// on update, this entire thing will be recreated and restarted
+type crdManager struct {
+	opts   component.Options
+	logger log.Logger
+	config *Config
+
+	discovery        *discovery.Manager
+	scraper          *scrape.Manager
+	cg               configGenerator
+	discoveryConfigs map[string]discovery.Configs
+	scrapeConfigs    map[string]*config.ScrapeConfig
+
+	mut sync.Mutex
+}
```

Comment:
I'd prefer to see the crdManager stuff in a separate file to make things a little more separated.